### PR TITLE
add support for __experimentalSkiSerialization to the shadow attribute

### PR DESCRIPTION
**Pull request by colinduwe**
**What?**
This is a backport of the merged Gutenberg PR https://github.com/WordPress/gutenberg/pull/59887
When a dynamic block defines __experimentalSkipSerialization in its block.supports.shadow the styles continue to be printed to the block wrapper element. This PR corrects that behavior so that shadow behaves like border, color, and others.
**Why?**
This provides the expected behavior for dynamic blocks that need to opt out of having the shadow styles printed to the block wrapper.
**How?**
Added a check at the start of the render function to return an empty array if the block has set __experimentalSkipSerialization

(This is my first attempt at submitting a Gutenberg PR to be included back into WP Core so welcome any advice if I'm doing this wrong)

Trac ticket: https://core.trac.wordpress.org/ticket/60784#ticket

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for backporting this fix @colinduwe 👍 

The change here matches https://github.com/WordPress/gutenberg/pull/59887.

I was going to note that we should add a unit test for this but it appears no tests have been added for the shadow block support like there are for typography, border etc.

@madhusudhand are you aware of any PRs in progress adding unit tests for the shadow support?

While this does allow support for an experimental block.json property (that should be stabilized soon), it does make the shadow block support consistent with all others while only introducing a call to a stable core function. I'd like to see this land.

**Comment by madhusudhand** [comment]
> @madhusudhand are you aware of any PRs in progress adding unit tests for the shadow support?

I have created https://github.com/WordPress/gutenberg/pull/60063 which tests these changes.

**Comment by vcanales** [comment]
👋 @colinduwe @madhusudhand what is the status here?

It looks like https://github.com/WordPress/gutenberg/pull/60063 has been approved. Are you facing any blockers?

**Comment by colinduwe** [comment]
@vcanales I think it is probably only blocked by my ignorance of the process here. I thought this was merely awaiting approval from someone with commit access.

**Comment by aaronrobertshaw** [comment]
> It looks like https://github.com/WordPress/gutenberg/pull/60063 has been approved. Are you facing any blockers?

Once https://github.com/WordPress/gutenberg/pull/60063 is merged, this should probably also include the unit test the code change relates to.

**Comment by madhusudhand** [comment]
@colinduwe could you bring the unit tests from https://github.com/WordPress/gutenberg/pull/60063 into this?

**Comment by colinduwe** [comment]
@madhusudhand I think I'm out of my depth here. I'm not sure how to do that.

**Comment by madhusudhand** [comment]
@colinduwe I have created #6613 with the unit tests. 
Please checkout #6613 to a branch and cherry pick the commit 8a4659608ba43363b4bd63fd76e99635c55141c8 into this PR

```
git fetch origin pull/6613/head:shadow-unit-tests
git cherry-pick 8a4659608ba43363b4bd63fd76e99635c55141c8
```

**Comment by colinduwe** [comment]
Your're a hero @madhusudhand. Thanks. Done.

File: tests/phpunit/tests/block-supports/shadow.php
Comment (vcanales) [inline tests/phpunit/tests/block-supports/shadow.php]: Thanks for bringing in the tests!

We have to rename the functions, since the `gutenberg_` prefix is meant to be used within Gutenberg and not core (https://github.com/WordPress/wordpress-develop/pull/6279/files#diff-a903d994435db8240aed96e253cfb2097ecbe9705d6f88ecdfd441422ecf2da3R52)

```suggestion
	public function test_wp_apply_shadow_support( $support, $value, $expected ) {
		$block_type  = self::register_shadow_block_with_support(
			'test/shadow-block',
			array( 'shadow' => $support )
		);
		$block_attrs = array( 'style' => array( 'shadow' => $value ) );
		$actual      = wp_apply_shadow_support( $block_type, $block_attrs );
```

Comment (madhusudhand) [reply tests/phpunit/tests/block-supports/shadow.php]: Fixed [here](https://github.com/WordPress/wordpress-develop/pull/6613/commits/5383012d66abb19d39388750fd3e1b46ec0d60b6). 

@colinduwe please pull and 
`git cherry-pick 5383012d66abb19d39388750fd3e1b46ec0d60b6`

Comment (colinduwe) [reply tests/phpunit/tests/block-supports/shadow.php]: Done

File: src/wp-includes/block-supports/shadow.php
Comment (tellthemachines) [inline src/wp-includes/block-supports/shadow.php:56]: Would be good to add a `since ` note to the docblock above; something like

` @since 6.6.0 Return early if __experimentalSkipSerialization is true.`

(it should be added just below `@since 6.3.0`)

**Review by tellthemachines — APPROVED** [review]
Thanks for updating @colinduwe, LGTM!

**Comment by tellthemachines** [comment]
Committed in [r58312](https://core.trac.wordpress.org/changeset/58312).

File: tests/phpunit/tests/block-supports/shadow.php
Comment (mukeshpanchal27) [inline tests/phpunit/tests/block-supports/shadow.php:60]: @tellthemachines why the ticket number is not mention for test?

Comment (tellthemachines) [reply tests/phpunit/tests/block-supports/shadow.php:60]: Thanks for spotting that! Fixed in https://core.trac.wordpress.org/changeset/58315.

**Pull request closed without merging**
